src/controller_examples/rabbitmq_controller/proof/helper_invariants/unchangeable.rs

```diff
@@ -0,0 +1,265 @@
+// Copyright 2022 VMware, Inc.
+// SPDX-License-Identifier: MIT
+#![allow(unused_imports)]
+use crate::external_api::spec::{EmptyAPI, EmptyTypeView};
+use crate::kubernetes_api_objects::{
+    api_method::*, common::*, dynamic::*, error::*, owner_reference::*, prelude::*, resource::*,
+};
+use crate::kubernetes_cluster::spec::{
+    cluster::*,
+    cluster_state_machine::Step,
+    controller::common::{ControllerActionInput, ControllerStep},
+    message::*,
+};
+use crate::rabbitmq_controller::{
+    common::*,
+    proof::{
+        helper_invariants::{owner_ref::*, predicate::*, proof::*, validation::*},
+        predicate::*,
+        resource::*,
+    },
+    spec::{reconciler::*, resource::*, types::*},
+};
+use crate::reconciler::spec::{reconciler::*, resource_builder::*};
+use crate::temporal_logic::{defs::*, rules::*};
+use crate::vstd_ext::{multiset_lib, seq_lib, string_view::*};
+use vstd::{multiset::*, prelude::*, string::*};
+
+verus! {
+
+pub open spec fn object_in_every_create_request_msg_satisfies_unchangeable(sub_resource: SubResource, rabbitmq: RabbitmqClusterView) -> StatePred<RMQCluster> {
+    let resource_key = get_request(sub_resource, rabbitmq).key;
+    |s: RMQCluster| {
+        forall |msg: RMQMessage|
+            #[trigger] s.in_flight().contains(msg)
+            && resource_create_request_msg(resource_key)(msg)
+            ==> unchangeable(sub_resource, msg.content.get_create_request().obj, rabbitmq)
+    }
+}
+
+pub open spec fn object_in_every_update_request_msg_satisfies_unchangeable(sub_resource: SubResource, rabbitmq: RabbitmqClusterView) -> StatePred<RMQCluster> {
+    let resource_key = get_request(sub_resource, rabbitmq).key;
+    |s: RMQCluster| {
+        forall |msg: RMQMessage|
+            #[trigger] s.in_flight().contains(msg)
+            && resource_update_request_msg(resource_key)(msg)
+            && s.resources().contains_key(resource_key)
+            && s.resources()[resource_key].metadata.resource_version == msg.content.get_update_request().obj.metadata.resource_version
+            ==> unchangeable(sub_resource, msg.content.get_update_request().obj, rabbitmq)
+    }
+}
+
+proof fn lemma_always_object_in_every_create_request_msg_satisfies_unchangeable(
```

Comment:
Can you also document this lemma and the one for update request? My understanding is that since the update request only updates part of the object, but our liveness property talks about the entire `.spec`, so you need some invariant to say the part, which never gets updated once created, remains unchanged. Is that correct?